### DISCUSSION TITLE
Support specifying composer options

### DISCUSSION THREAD

**Discussion by hussainweb** [renovatebot/renovate#24899] [Suggest an Idea]
**What would you like Renovate to be able to do?**
I'd like to be able to specify composer options like what flags to pass and what environment variables to set. For example, in [this comment](https://github.com/renovatebot/renovate/pull/6268#issuecomment-631430423), @JamieMagee rightly says he is nervous to give composer unlimited memory. But unfortunately, this is typically required for a complex project typical in Drupal.

Another thing I'd want to specify is if `--update-with-dependencies` or `--update-with-all-dependencies` should be passed. Right now, it's only `--update-with-dependencies` and there is no way if I want for me to specify otherwise.

I am thinking all of this should be made configurable.

**Describe the solution you'd like**
I am seeing there is [a way to specify configuration for managers](https://docs.renovatebot.com/modules/manager/#file-matching). I'd like to add these options there. Specifying this in `packageRules` is another idea.

**Describe alternatives you've considered**
I can't think of any other way to achieve this currently with Renovate.

**Additional context**
N/A

**Comment by hussainweb** [comment]
I'm happy to work on a PR. I don't know where to start writing code for specifying additional options. Can you help here please @rarkins @JamieMagee?

**Comment by rarkins** [comment]
There's a few ways to think about this:

1. Are any of the options potentially "dangerous" to the bot owner? e.g. could they let a repository user add a configuration that causes the bot to crash?

2. Does Renovate need to understand all the options individually, or could it just have a dumb configuration option like an `artifactOptions` string array and then it's up to the artifacts code to know what those options mean?

**Comment by hussainweb** [comment]
Thank you, @rarkins. Here are my thoughts:

1. I don't think so. It is possible `COMPOSER_MEMORY_LIMIT` could be considered borderline but it only impacts `composer`. The other option also impacts `composer` and has no impact on code outside the repository.

2. I am not sure what you mean by "artifacts" here. If you mean that `composer` is an artefact, then that makes sense. If you mean the files it generates (composer.json and composer.lock), then maybe these options don't make sense there. These options affect how artefacts are going to be generated, not the artefacts themselves. Does that make sense?

FWIW, I thought of a workaround for the first one but that is on the bot level. I have yet to try it but if I run `renovate` inside an environment which already has `COMPOSER_MEMORY_LIMIT` set, then that would cascade down to `composer` as well, right?

**Comment by rarkins** [comment]
> 2\. I am not sure what you mean by "artifacts" here. If you mean that `composer` is an artefact, then that makes sense. If you mean the files it generates (composer.json and composer.lock), then maybe these options don't make sense there. These options affect how artefacts are going to be generated, not the artefacts themselves. Does that make sense?

When we use a tool like `composer` to update lock files or other files, we call that "artifact" generation. So I mean, can Renovate just blindly pass these options through if we let a user configure them, or does it need to understand or filter them?

> FWIW, I thought of a workaround for the first one but that is on the bot level. I have yet to try it but if I run `renovate` inside an environment which already has `COMPOSER_MEMORY_LIMIT` set, then that would cascade down to `composer` as well, right?

Only if you set `trustLevel=true`. Otherwise Renovate does not pass the full environment to child processes.

Perhaps start this way. Can you list the top 5 options you'd like to pass to `composer` and whether you think they are administrator-only options (configurable by the bot admin only) or user options (configurable by the repository owner)? e.g. I think the composer memory limit should be admin-only.

**Comment by hussainweb** [comment]
> can Renovate just blindly pass these options through if we let a user configure them, or does it need to understand or filter them?

Renovate can blindly pass these options, but there could be some difference in verbiage. For example, the config option name might be `updateWithAllDependencies` but the parameter passed to the composer command would be `--with-all-dependencies` instead of the current `--with-dependencies`.

I suppose the other option for the memory limit is slightly different. Here, you might not specify an option but set an environment variable like so:
```
       extraEnv: {
         COMPOSER_CACHE_DIR: cacheDir,
         COMPOSER_MEMORY_LIMIT: -1,
       },
```

> Can you list the top 5 options you'd like to pass to composer

I can only think of these two which are relevant to `composer update`.
1. Memory limit (per-repo or admin-only)
2. Updating all dependencies (per-repo)

**Comment by hussainweb** [comment]
I suppose there could be more options but I have never really had to use any of these and so I am not sure if they are really valuable. I don't see people using this in an automated way in any case. Still, these are just options you pass to `composer` and you don't really need to worry about much.

1. --root-reqs
2. --prefer-stable
3. --prefer-lowest

Another thing I thought of. I see Renovate uses a Docker image to run `composer`. Is it updated? Is there a way to specify an alternate composer version just in case? For example, the current version is 1.10.6 but I don't know what version is in the image used by Renovate.

**Comment by pedrommone** [comment]
@hussainweb currently we're stuck with lock file maintenance because all our projects are big and need changing the memory limit to work with.

Is there a workaround for fixing that using environment variables?

**Comment by hussainweb** [comment]
@pedrommone, maybe I am not getting the context because the issue is old. If you want to change the memory limit available to composer, you can set [`COMPOSER_MEMORY_LIMIT`](https://getcomposer.org/doc/articles/troubleshooting.md#memory-limit-errors) environment variable for the bot, but you also need to set [`trustLevel=true`](https://docs.renovatebot.com/self-hosted-configuration/#trustlevel) so that the environment variable is actually made available to composer.

**Comment by mikepsinn** [comment]
> @pedrommone, maybe I am not getting the context because the issue is old. If you want to change the memory limit available to composer, you can set [`COMPOSER_MEMORY_LIMIT`](https://getcomposer.org/doc/articles/troubleshooting.md#memory-limit-errors) environment variable for the bot, but you also need to set [`trustLevel=true`](https://docs.renovatebot.com/self-hosted-configuration/#trustlevel) so that the environment variable is actually made available to composer.

I don't see  trustLevel in the docs: https://docs.renovatebot.com/self-hosted-configuration/#trustlevel

Is that comment still accurate?

Is that applicable to non-self-hosted situations?

Does anyone have an example `renovate.json` where they've solved this issue?

**Comment by JamieMagee** [comment]
@mikepsinn `trustLevel` was split into 3 separate variables in #9388. What you're looking for is now called [`exposeAllEnv`](https://docs.renovatebot.com/self-hosted-configuration/#exposeallenv). That should allow you to set `COMPOSER_MEMORY_LIMIT` on your own host, and Renovate will pick up that environment variable and use it.

**Comment by viceice** [comment]
Alternative is to use `customEnvVariables` to only override one variable.

https://docs.renovatebot.com/self-hosted-configuration/#customenvvariables

**Comment by felixranesberger** [comment]
I would like to implement this feature, since I need it myself, in order to get my Renovate Gitlab-CE setup working. 

I already started working on a pull request, but I'm not sure on how to test my changes. Is there a documentation for local development?

Here are my current changes: https://github.com/felixranesberger/renovate/commit/67c7cc08b046bfbaba1822afa984cd6bb7a925f2

**Comment by rarkins** [comment]
https://github.com/renovatebot/renovate/blob/main/docs/development/local-development.md
